File: spring-web-java/src/main/java/spring/web/java/domain/auth/service/AuthService.java
package spring.web.java.domain.auth.service;

import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import lombok.RequiredArgsConstructor;
import spring.web.java.domain.member.Member;
import spring.web.java.domain.member.dto.MemberLoginRequest;
import spring.web.java.domain.member.repository.MemberRepository;
import spring.web.java.domain.token.Token;
import spring.web.java.domain.token.dto.TokenRequest;
import spring.web.java.domain.token.dto.TokenResponse;
import spring.web.java.domain.token.repository.TokenRepository;
import spring.web.java.global.common.JwtTokenProvider;
import spring.web.java.global.common.ResponseMessage;
import spring.web.java.global.exception.DomainException;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class AuthService {

	private final JwtTokenProvider jwtTokenProvider;
	private final TokenRepository tokenRepository;
	private final MemberRepository memberRepository;
	private final PasswordEncoder passwordEncoder;

	@Transactional
	public TokenResponse login(MemberLoginRequest memberLoginRequest) {
		Member member = memberRepository.findByAccount(memberLoginRequest.account())
			.orElseThrow(() -> new DomainException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED));

		if (!passwordEncoder.matches(memberLoginRequest.password(), member.getPassword())) {
			throw new DomainException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED);
		}

		String accessToken = jwtTokenProvider.createAccessToken(member.getId(), member.getRole());
		String refreshToken = jwtTokenProvider.createRefreshToken();

		tokenRepository.save(Token.create(member.getId(), refreshToken));

		return new TokenResponse(accessToken, refreshToken);
	}

	public TokenResponse refreshToken(TokenRequest tokenRequest) {
		Claims claims;

		try {
			claims = jwtTokenProvider.parseAccessToken(tokenRequest.accessToken());
		} catch (ExpiredJwtException e) {
			claims = e.getClaims();
		}

		jwtTokenProvider.parseRefreshToken(tokenRequest.refreshToken());

		Long memberId = Long.valueOf(String.valueOf(claims.get("memberId")));

		String refreshToken = tokenRepository.findById(memberId)
			.map(Token::getRefreshToken)
			.filter(token -> tokenRequest.refreshToken().equals(token))
			.orElseThrow(() -> new DomainException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED));

		Member member = memberRepository.findById(memberId)
			.orElseThrow(() -> new DomainException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED));

		return new TokenResponse(jwtTokenProvider.createAccessToken(member.getId(), member.getRole()), refreshToken);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/auth/service/AuthService.kt
package spring.web.kotlin.domain.auth.service

import io.jsonwebtoken.ExpiredJwtException
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpStatus
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import spring.web.kotlin.domain.member.dto.MemberLoginRequest
import spring.web.kotlin.domain.member.repository.MemberRepository
import spring.web.kotlin.domain.token.Token
import spring.web.kotlin.domain.token.dto.TokenRequest
import spring.web.kotlin.domain.token.dto.TokenResponse
import spring.web.kotlin.domain.token.repository.TokenRepository
import spring.web.kotlin.global.common.ResponseMessage
import spring.web.kotlin.global.config.JwtTokenProvider
import spring.web.kotlin.global.exception.BaseException

@Service
@Transactional(readOnly = true)
class AuthService(
    private val jwtTokenProvider: JwtTokenProvider,
    private val tokenRepository: TokenRepository,
    private val memberRepository: MemberRepository,
    private val passwordEncoder: PasswordEncoder
) {

    @Transactional
    fun login(memberLoginRequest: MemberLoginRequest): TokenResponse {
        val member = (memberRepository.findByAccount(memberLoginRequest.account)
            ?: throw BaseException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED))

        if (!passwordEncoder.matches(memberLoginRequest.password, member.password)) {
            throw BaseException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED)
        }

        val accessToken = jwtTokenProvider.createAccessToken(member.id!!, member.role)
        val refreshToken = jwtTokenProvider.createRefreshToken()

        tokenRepository.save(Token.create(member.id!!, refreshToken))

        return TokenResponse(accessToken, refreshToken)
    }

    fun refreshToken(tokenRequest: TokenRequest): TokenResponse {
        val claims = runCatching { jwtTokenProvider.parseAccessToken(tokenRequest.accessToken) }
            .getOrElse { throwable ->
                when (throwable) {
                    is ExpiredJwtException -> throwable.claims
                    else -> throw throwable
                }
            }

        jwtTokenProvider.parseRefreshToken(tokenRequest.refreshToken)

        val memberId = claims["memberId"] as Long

        val token = tokenRepository.findByIdOrNull(memberId)
            ?: throw BaseException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED)

        if (token.refreshToken != tokenRequest.refreshToken) {
            throw BaseException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED)
        }

        val member = memberRepository.findByIdOrNull(memberId)
            ?: throw BaseException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED)

        return TokenResponse(jwtTokenProvider.createAccessToken(member.id!!, member.role), token.refreshToken)
    }
}